archive/brotli.cpp

```diff
@@ -0,0 +1,101 @@
+// SPDX-FileCopyrightText: 2024 David Zero <zero-one@zer0-one.net>
+//
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "archive/brotli.h"
+
+#include "brotli/decode.h"
+#include <tl/expected.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <span>
+#include <string_view>
+#include <vector>
+
+namespace archive {
+
+std::string_view to_string(BrotliError err) {
+    switch (err) {
+        case BrotliError::DecoderState:
+            return "Failed to create brotli decoder state";
+        case BrotliError::InputCorrupt:
+            return "Input is corrupt or truncated";
+        case BrotliError::InputEmpty:
+            return "Input is empty";
+        case BrotliError::MaximumOutputLengthExceeded:
+            return "Output buffer exceeded maximum allowed length";
+        case BrotliError::BrotliInternalError:
+            return "Decode failure";
+    }
+
+    return "Unknown error";
+}
+
+tl::expected<std::vector<std::byte>, BrotliError> brotli_decode(std::span<std::byte const> const input) {
+    if (input.empty()) {
+        return tl::unexpected{BrotliError::InputEmpty};
+    }
+
+    std::unique_ptr<BrotliDecoderState, decltype(&BrotliDecoderDestroyInstance)> br_state(
+            BrotliDecoderCreateInstance(nullptr, nullptr, nullptr), BrotliDecoderDestroyInstance);
+
+    if (br_state == nullptr) {
+        return tl::unexpected{BrotliError::DecoderState};
+    }
+
+    // Cap output buffer at 1GB. If we hit this, something fishy is probably
+    // going on, and we should bail before we OOM.
+    std::size_t constexpr kMaxOutSize = 1000000000;
```

Comment:
This isn't 1GB. But it's also seems totally reasonable to have the output >= 1GB.

Reply:
1. That is definitely 1GB. If it's not, then I need you to explain how.
2. I can't imagine a scenario where you legitimately need to decode a 1GB of content from a single HTTP response, so I need you to explain this one too.

Reply:
Bah! you're correct, I got GiB and GB confused... again :<

Reply:
2) 
this is terminal output from seconds ago
```
grayhatter@host ~ $ rsync -vPz ~/[redacted/server-local.zip gr.ht:/srv/http/gr.ht/[redacted].zip
[redacted]-server-local.zip
    7,322,472,876 100%    5.01MB/s    0:01:01 (xfr#1, to-chk=0/1)

sent 7,320,562,208 bytes  received 225,626 bytes  5,050,202.11 bytes/sec
total size is 7,322,472,876  speedup is 1.01